instrument/src/Instrument/Client.hs

```diff
@@ -349,13 +349,17 @@ getSamplers ss = M.toList `fmap` readIORef ss
 -- return it.
 getRef :: Ord k => IO b -> k -> IORef (M.Map k b) -> IO b
 getRef f name mapRef = do
-  empty <- f
-  atomicModifyIORef mapRef $ \m ->
-    case M.lookup name m of
-      Nothing ->
-        let m' = M.insert name empty m
-         in (m', empty)
-      Just ref -> (m, ref)
+  mapRef' <- readIORef mapRef
```

Comment:
I'm not sure if it's significant but I think this might be slightly semantically different. Previously we put an atomic write lock around the whole read-or-insert. If the name already existed in a map then we didn't actually have to write so that might have been wasteful. But in this new implementation, isn't it possible for the map to be modified between the `readIORef` and the `atomicModifyIORef` where it wasn't possible before?`

Reply:
This map is append-only, and we don't modify values once they are inserted, so this is safe. This essentially implements [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking). The only possibility is that another thread inserts into the map in between of first `readIORef` and `atomicModifyIORef` – that's why we perform lookup again inside `atomicModifyIORef`.